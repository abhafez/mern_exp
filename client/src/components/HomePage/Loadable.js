import React from 'react';
import loadable from 'utils/loadable';
import { LoaderExampleText } from 'components/LoadingIndicator';

export default loadable(() => import('./index'), {
  fallback: <LoaderExampleText />,
});
