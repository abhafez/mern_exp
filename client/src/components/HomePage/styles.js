import styled from 'styled-components';

export const PromoFilterstyled = styled.div`
  display: inline;
  margin-right: 1em;
`;

export const DepartmentFilterstyled = styled.div`
  display: inline;
  margin-rigth: 1em;
`;
