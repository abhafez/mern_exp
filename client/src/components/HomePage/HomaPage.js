import React, { useEffect, useState } from 'react';
import { Message, Segment, Divider, Button, Icon } from 'semantic-ui-react';
import useRequest, { get } from 'hooks/useRequest';
import { Shimmer } from 'components/shared/Shimmer';
import ProductsTable from 'components/ProductsTable';
import FilterDropdown from 'components/FilterDropdown/FilterDropdown';
import Search from 'components/Search';
import { promos, departments } from 'constants/filters';
import { DepartmentFilterstyled, PromoFilterstyled } from './styles';

function HomePage() {
  const [productList, setProductList] = useState(null);
  const [maxLength, setMaxLength] = useState(null);
  const [itemsPerPage, setItemPerPage] = useState(8);
  const [pageNumber, setPageNumber] = useState(1);
  const [departmentSelection, setDepartmentSelection] = useState();
  const [promoSelection, setPromoSelection] = useState();
  const [productName, setProductName] = useState('');
  const [submitName, setSubmitName] = useState(false);

  const { response, setRequest, error } = useRequest(
    get,
    `products?page=${pageNumber}&count=${itemsPerPage}${
      promoSelection ? `&promo=${promoSelection}` : ''
    }${departmentSelection ? `&department=${departmentSelection}` : ''}${
      productName ? `&name=${productName}` : ''
    }`,
  );

  const promoFilter = (e, { value }) => setPromoSelection(value);
  const departmentFilter = (e, { value }) => setDepartmentSelection(value);
  const handleNameInput = (e) => setProductName(e.target.value);
  const searchByName = () => setSubmitName(true);
  const clear = () => {
    setDepartmentSelection(null);
    setProductName('');
    setPromoSelection(null);
  };

  useEffect(() => {
    setRequest(true);
  }, []);

  useEffect(() => {
    if (submitName) {
      setRequest(true);
      setSubmitName(false);
    }
  }, [submitName]);

  useEffect(() => {
    if (pageNumber || itemsPerPage || departmentSelection || promoSelection) {
      setRequest(true);
    }
  }, [pageNumber, itemsPerPage, departmentSelection, promoSelection]);

  const handlItemsPerPage = (e, data) => {
    setItemPerPage(data.value);
    setPageNumber(1);
  };

  useEffect(() => {
    if (response && response.data) {
      setProductList(response.data.list);
      setMaxLength(response.data.filteredLength);
    }
  }, [response]);
  const tableHeaders = ['name', 'price', 'code', 'department'];

  return (
    <>
      <Segment basic>
        <PromoFilterstyled>
          Filters:{' '}
          <FilterDropdown
            value={promoSelection}
            options={promos}
            onFilter={promoFilter}
          />
        </PromoFilterstyled>
        <DepartmentFilterstyled>
          <FilterDropdown
            options={departments}
            value={departmentSelection}
            onFilter={departmentFilter}
          />
        </DepartmentFilterstyled>

        <Divider hidden />
        <Search
          getName={handleNameInput}
          searchByName={searchByName}
          value={productName}
        />
        <Button animated="vertical" onClick={clear}>
          <Button.Content hidden>Clear</Button.Content>
          <Button.Content visible>
            <Icon name="trash" />
          </Button.Content>
        </Button>
      </Segment>
      {!productList && <Shimmer />}
      {error && (
        <Message negative>
          <Message.Header>Something went wrong</Message.Header>
        </Message>
      )}
      {productList && (
        <ProductsTable
          tableHeaders={tableHeaders}
          productList={productList}
          itemsPerPage={itemsPerPage}
          handlItemsPerPage={handlItemsPerPage}
          maxLength={maxLength}
          pageNumber={pageNumber}
          setPageNumber={setPageNumber}
        />
      )}
    </>
  );
}

export default HomePage;
