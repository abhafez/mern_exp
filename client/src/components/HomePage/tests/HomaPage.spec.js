import React from 'react';
import { render } from '@testing-library/react';
import HomePage from '../HomaPage';

describe('<Homepage />', () => {
  it('Contains Match snapshot', () => {
    const { container } = render(<HomePage />);
    expect(container).toMatchSnapshot();
  });
});
