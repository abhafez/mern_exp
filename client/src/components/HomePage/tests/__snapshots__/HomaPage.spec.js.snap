// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Homepage /> Contains Match snapshot 1`] = `
<div>
  <div
    class="ui basic segment"
  >
    <div
      class="sc-AxirZ fLHxDp"
    >
      Filters:
       
      <div
        aria-expanded="false"
        class="ui search selection dropdown"
        role="combobox"
        testid="dropdown"
      >
        <input
          aria-autocomplete="list"
          autocomplete="off"
          class="search"
          tabindex="0"
          type="text"
          value=""
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          class="text"
          role="alert"
        />
        <i
          aria-hidden="true"
          class="dropdown icon"
        />
        <div
          class="menu transition"
          role="listbox"
        >
          <div
            aria-checked="false"
            aria-selected="true"
            class="selected item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Toy-20
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Game-30
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Gift-50
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Book-80
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-AxiKw ggyaPd"
    >
      <div
        aria-expanded="false"
        class="ui search selection dropdown"
        role="combobox"
        testid="dropdown"
      >
        <input
          aria-autocomplete="list"
          autocomplete="off"
          class="search"
          tabindex="0"
          type="text"
          value=""
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          class="text"
          role="alert"
        />
        <i
          aria-hidden="true"
          class="dropdown icon"
        />
        <div
          class="menu transition"
          role="listbox"
        >
          <div
            aria-checked="false"
            aria-selected="true"
            class="selected item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Gaming
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Toys
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Gifts
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Books
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ui hidden divider"
    />
    <input
      class="sc-AxjAm iMmbHG"
      id="name-search"
      placeholder="Product Name"
      value=""
    />
    <button
      class="ui button"
      data-testid="btn"
      id="search-btn"
      type="submit"
    >
      Search by Name
    </button>
    <button
      class="ui vertical animated button"
    >
      <div
        class="hidden content"
      >
        Clear
      </div>
      <div
        class="visible content"
      >
        <i
          aria-hidden="true"
          class="trash icon"
        />
      </div>
    </button>
  </div>
  <div
    class="ui placeholder"
  >
    <div
      class="paragraph"
    >
      <div
        class="line"
      />
      <div
        class="line"
      />
      <div
        class="line"
      />
      <div
        class="line"
      />
      <div
        class="line"
      />
      <div
        class="line"
      />
      <div
        class="line"
      />
    </div>
  </div>
</div>
`;
